src/Microsoft.DotNet.ProjectModel/WorkspaceContext.cs

```diff
@@ -0,0 +1,427 @@
+        /// are added to the WorkspaceContext.
+        /// 
+        /// If the given path points to a project.json, all the projects it referenced as well as itself
+        /// are added to the WorkspaceContext.
+        /// </summary>
+        public static WorkspaceContext CreateFrom(string projectPath, string configuration)
+        {
+            var projectPaths = ResolveProjectPath(projectPath);
+            if (projectPaths == null || !projectPaths.Any())
+            {
+                return null;
+            }
+
+            var context = new WorkspaceContext(projectPaths, configuration);
+            return context;
+        }
+
+        public static WorkspaceContext CreateFrom(string projectPath)
+        {
+            return CreateFrom(projectPath, "Debug");
+        }
+
+        public void AddProject(string path)
+        {
+            var projectPath = NormalizeProjectPath(path);
+
+            if (projectPath != null)
+            {
+                _needRefresh = _projects.Add(path);
+            }
+        }
+
+        public void RemoveProject(string path)
+        {
+            _needRefresh = _projects.Remove(path);
+        }
+
+        public IReadOnlyList<string> GetAllProjects()
+        {
+            Refresh();
+            return _projects.ToList().AsReadOnly();
+        }
+
+        /// <summary>
+        /// Refresh the WorkspaceContext to update projects collection
+        /// </summary>
+        public void Refresh()
+        {
+            if (!_needRefresh)
+            {
+                return;
+            }
+
+            var basePaths = new List<string>(_projects);
+            _projects.Clear();
+
+            foreach (var projectDirectory in basePaths)
+            {
+                var project = GetProject(projectDirectory);
+                if (project == null)
+                {
+                    continue;
+                }
+
+                _projects.Add(project.ProjectDirectory);
+
+                foreach (var projectContext in GetProjectContexts(project.ProjectDirectory))
+                {
+                    foreach (var reference in GetProjectReferences(projectContext))
+                    {
+                        var referencedProject = GetProject(reference.Path);
+                        if (referencedProject != null)
+                        {
+                            _projects.Add(referencedProject.ProjectDirectory);
+                        }
+                    }
+                }
+            }
+
+            _needRefresh = false;
+        }
+
+        public IReadOnlyList<ProjectContext> GetProjectContexts(string projectPath)
+        {
+            return _projectContextsCache.AddOrUpdate(
+                projectPath,
+                key => AddProjectContextEntry(key, null),
+                (key, oldEntry) => AddProjectContextEntry(key, oldEntry)).ProjectContexts;
+        }
+
+        private Project GetProject(string projectDirectory)
```

Comment:
No, that won't happen. The projects cache here is used in ProjectContext for resolving the projects. The idea is to pool the projects so that project.json are not read repeatedly when multiple ProjectContexts are created. See here: https://github.com/dotnet/cli/pull/277/files#diff-d32cd079f5a7a3f2ff9e2f7ab2487413R237